Support partial failures in multirun mode.

jieru-hu opened the pull request:
This is to address #1377

This will not change the behavior of RUN but will change the behavior of MULTIRUN: sweepers will finish all the jobs despite any single Job failures. 
This PR only address BasicSweeper as a starting point.

omry commented on tests/test_basic_sweeper.py:
```diff
@@ -54,3 +58,42 @@ def test_split(
+        "+divisor=1,0,2",
+        "hydra.run.dir=" + str(tmpdir),
+    ]
+    with Popen(cmd, stdout=PIPE, stderr=PIPE) as p:
+        stdout, stderr = p.communicate()
+        err = stderr.decode("utf-8").rstrip().replace("\r\n", "\n")
+        out = stdout.decode("utf-8").rstrip().replace("\r\n", "\n")
+        assert p.returncode != 0
+
```
Add a comment explaining why you are not using run_python_script(), run_process or run_with_error().

omry commented on tests/test_basic_sweeper.py:
```diff
@@ -54,3 +58,42 @@ def test_split(
+    monkeypatch: Any,
+    tmpdir: Any,
+) -> None:
+    monkeypatch.chdir("tests/test_apps/app_can_fail")
```
since you are overriding the run_dir, I think you don't have to patch the chdir.
just run it from the Hydra root dir with full path.

omry reviewed (changes requested):
This will break existing sweepers, right?
The result can now suddenly be an exception object.
See my inline comment about a getter method.

jieru-hu commented:
>Can this fit in one line (move the comment above it).
I think it would be good to add an getter for the result that would raise an exception if the result is an exception.
This can be used as default behavior in sweepers (unless they want to have a modified behavior).

Thanks, that sounds better! I've made return_value a property.

>Why not
print(f"{val=}")

That's what I did in the first place, but python <3.8 [does not like that syntax](https://app.circleci.com/pipelines/github/facebookresearch/hydra/8206/workflows/518e869a-303f-48e1-bbeb-bdcae932f59b/jobs/69686)

>Add a comment explaining why you are not using run_python_script(), run_process or run_with_error().

Added an optional params to `run_process` instead. 

>since you are overriding the run_dir, I think you don't have to patch the chdir.
just run it from the Hydra root dir with full path.

thanks! fixed!

jieru-hu commented:
forgot to put this back to ready for review :)

omry commented on hydra/_internal/core_plugins/basic_sweeper.py:
```diff
@@ -152,8 +153,21 @@ def sweep(self, arguments: List[str]) -> Any:
+
             initial_job_idx += len(batch)
             returns.append(results)
+        if job_failed:
+            raise ValueError("Application multirun failed: not all jobs succeeded.")
```
1. Use a subclass of HydraException or CompactHydraException.
See the code to learn about the difference.

2. We should probably chain the first exception to the raised one to enable a stack trace. This might require changes to how exceptions are handled in run_and_report().

jieru-hu replied:
Thanks for the pointers. I've update the exception to be `CompactHydraException` and chained the first exception. 

Here is what the stack trace looks like  as of (10f857d4380c0a89d19f3b6a915430a283863814)

### Without HYDRA_FULL_ERROR=1
```
$  python tests/test_apps/app_can_fail/my_app.py +divisor=1,0,0 -m
[2021-03-17 11:27:35,688][HYDRA] Launching 3 jobs locally
[2021-03-17 11:27:35,689][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-17 11:27:35,762][HYDRA]        #1 : +divisor=0
[2021-03-17 11:27:35,846][HYDRA]        #2 : +divisor=0
[2021-03-17 11:27:35,921][HYDRA] Job #1 failed.
[2021-03-17 11:27:35,921][HYDRA] Job #2 failed.
hydra.errors.FailedMultirunException: Multirun failed, failed job: #1,#2

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```


### With HYDRA_FULL_ERROR=1
```
$ HYDRA_FULL_ERROR=1  python tests/test_apps/app_can_fail/my_app.py +divisor=1,0,0 -m
[2021-03-17 11:25:40,050][HYDRA] Launching 3 jobs locally
[2021-03-17 11:25:40,050][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-17 11:25:40,133][HYDRA]        #1 : +divisor=0
[2021-03-17 11:25:40,208][HYDRA]        #2 : +divisor=0
[2021-03-17 11:25:40,283][HYDRA] Job #1 failed.
[2021-03-17 11:25:40,283][HYDRA] Job #2 failed.
hydra.errors.FailedMultirunException: Multirun failed, failed job: #1,#2

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 14, in <module>
    my_app()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/main.py", line 29, in decorated_main
    _run_hydra(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 369, in _run_hydra
    run_and_report(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 213, in run_and_report
    raise ex
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 210, in run_and_report
    return func()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 370, in <lambda>
    lambda: hydra.multirun(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/hydra.py", line 126, in multirun
    return sweeper.sweep(arguments=task_overrides)
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/core_plugins/basic_sweeper.py", line 174, in sweep
    raise first_exception from FailedMultirunException(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/core_plugins/basic_sweeper.py", line 162, in sweep
    _ = r.return_value
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 197, in return_value
    raise self._return_value
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 131, in run_job
    ret._return_value = task_function(task_cfg)
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero
```

omry replied:
This looks good. Can you verify that it looks similar when using a more sophisticated launcher? (not the basic one).

omry commented on hydra/_internal/hydra.py:
```diff
@@ -90,14 +91,19 @@ def run(
             configure_logging=with_log_configuration,
         )
 
+        if ret.job_result == JobResult.FAILED:
+            raise ret._return_value
```
can you show an example exception printout for this case?

omry replied:
Use `ret.return_value()`

Since it's a property, you can probably just replace this with 
```python
return ret.return_value
```

omry commented on hydra/core/utils.py:
```diff
@@ -166,14 +173,29 @@ def register(name: str, f: Any) -> None:
+
+    @property
+    def return_value(self) -> Any:
+        if self.job_result is None:
+            raise ValueError("job_result not yet available.")
```
This should never happen. an assertion makes more sense.
```suggestion
        assert self.job_result is not None, "job_result not yet available"
```

If not an assertion, then I think we should use some other exception type. ValueError does not seem right here.

omry commented on hydra/test_utils/test_utils.py:
```diff
@@ -377,6 +377,8 @@ def run_process(
     cmd: Any,
     env: Any = None,
     print_error: bool = True,
+    # set raise_exception to False if exception is expected and needs to be validated.
+    raise_exception: bool = True,
```
Use docstring to document parameters.

omry commented on hydra/_internal/core_plugins/basic_sweeper.py:
```diff
@@ -152,8 +154,26 @@ def sweep(self, arguments: List[str]) -> Any:
+
+            idx = initial_job_idx
+
+            for r in results:
+                try:
+                    _ = r.return_value
+                except Exception as e:
+                    job_failed.append(idx)
+                    log.error(f"Job #{idx} failed.")
+                    if first_exception is None:
+                        first_exception = e
+                idx += 1
+
             initial_job_idx += len(batch)
             returns.append(results)
+
+        if first_exception is not None:
+            raise first_exception from FailedMultirunException(
+                f"Multirun failed, failed job: {','.join(['#' + str(j) for j in job_failed])}"
+            )
```
I think that in order to keep the basic launcher basic, we should just let the first exception fly and terminate the loop.

just access the result to trigger the exception.

omry commented on hydra/core/utils.py:
```diff
@@ -166,14 +173,28 @@ def register(name: str, f: Any) -> None:
     working_dir: Optional[str] = None
     task_name: Optional[str] = None
+    job_result: Optional[JobResult] = None
+    _return_value: Any = None
+
+    @property
+    def return_value(self) -> Any:
+        assert self.job_result is not None, "job_result not yet available"
+        if self.job_result == JobResult.COMPLETED:
+            return self._return_value
+        else:
+            raise self._return_value
```
https://docs.python.org/3/library/functions.html#property

`@property` supports setters.  You can use it and keep the member with the same name as the property.

omry replied:
Raise the proper exception here.
```
raise HydraJobException("Job failed") from self._return_value.
```

This way you don't need to have the logic to raise it in many places.

omry commented on hydra/errors.py:
```diff
@@ -39,3 +39,7 @@ def __init__(
         self.options = options
+
+
+class FailedMultirunException(CompactHydraException):
+    ...
```
```suggestion
class HydraJobException(CompactHydraException):
    """Raised when accessing a result of a failed job"""
    ...
```

omry commented on tests/test_basic_sweeper.py:
```diff
@@ -54,3 +57,57 @@ def test_split(
     assert lret == expected
+
+
+def test_partial_failure(
+    tmpdir: Any,
```
nice test :)

omry commented on hydra/_internal/hydra.py:
```diff
@@ -90,14 +91,19 @@ def run(
             configure_logging=with_log_configuration,
         )
 
+        if ret.job_result == JobResult.FAILED:
+            raise ret.return_value
```
```suggestion
        # access the result to trigger an exception in case the job failed.
        _ = ret.return_value 
```

jieru-hu commented:
This is what the stacktrace looks like as of 798363fbca05307d09a41f150c36c84638b504f7

### RUN
```
$  python tests/test_apps/app_can_fail/my_app.py +divisor=0
hydra.errors.HydraJobException: Job failed.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

### BasicSweeper multirun
```
$  python tests/test_apps/app_can_fail/my_app.py +divisor=1,0 -m
[2021-03-17 16:18:50,429][HYDRA] Launching 2 jobs locally
[2021-03-17 16:18:50,429][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-17 16:18:50,502][HYDRA]        #1 : +divisor=0
hydra.errors.HydraJobException: Job failed.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

```

### ax sweeper
```
$ python example/banana.py -m 
[2021-03-17 17:26:42,089][HYDRA] AxSweeper is optimizing the following parameters: 
banana.x: range=[-5, 5]
banana.y: range=[-5.0, 10.1]
[INFO 03-17 17:26:42] ax.service.utils.instantiation: Inferred value type of ParameterType.INT for parameter banana.x. If that is not the expected value type, you can explicity specify 'value_type' ('int', 'float', 'bool' or 'str') in parameter dict.
[INFO 03-17 17:26:42] ax.service.utils.instantiation: Inferred value type of ParameterType.FLOAT for parameter banana.y. If that is not the expected value type, you can explicity specify 'value_type' ('int', 'float', 'bool' or 'str') in parameter dict.
[INFO 03-17 17:26:42] ax.modelbridge.dispatch_utils: Using Bayesian Optimization generation strategy: GenerationStrategy(name='Sobol+GPEI', steps=[Sobol for 5 trials, GPEI for subsequent trials]). Iterations after 5 will take longer to generate due to  model-fitting.
[2021-03-17 17:26:42,334][HYDRA] AxSweeper is launching 5 jobs
[2021-03-17 17:26:42,964][HYDRA] Launching 5 jobs locally
[2021-03-17 17:26:42,964][HYDRA]        #0 : banana.x=-1 banana.y=8.733434039540589
[2021-03-17 17:26:43,152][HYDRA]        #1 : banana.x=1 banana.y=2.687994828354567
[2021-03-17 17:26:43,326][HYDRA]        #2 : banana.x=-5 banana.y=4.666555053927004
[2021-03-17 17:26:43,524][HYDRA]        #3 : banana.x=5 banana.y=2.683330000285059
[2021-03-17 17:26:43,720][HYDRA]        #4 : banana.x=-5 banana.y=-2.378906474262476
hydra.errors.HydraJobException: Job failed.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/plugins/hydra_ax_sweeper/example/banana.py", line 18, in banana
    z = a / 0
ZeroDivisionError: division by zero

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

```

jieru-hu commented on hydra/core/utils.py:
```diff
@@ -166,14 +174,32 @@ def register(name: str, f: Any) -> None:
+        if self.job_result == JobResult.COMPLETED:
+            return self._return_value
+        else:
+            raise self._return_value from HydraJobException("Job failed.")
```
fixed! :)

jieru-hu commented on tests/test_basic_sweeper.py:
```diff
@@ -54,3 +57,54 @@ def test_split(
+        """\
+        hydra.errors.HydraJobException: Job failed.
+
+        The above exception was the direct cause of the following exception:
```
Thanks!

After comparing several options, I've made this a `HydraException` with some edits to `run_and_report`

This is what the stack track looks like now

### run
```
$ python tests/test_apps/app_with_runtime_config_error/my_app.py 
Job failed.
Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_with_runtime_config_error/my_app.py", line 13, in my_app
    foo(cfg)
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_with_runtime_config_error/my_app.py", line 8, in foo
    cfg.foo = "bar"  # does not exist in the config
omegaconf.errors.ConfigAttributeError: Key 'foo' is not in struct
    full_key: foo
    object_type=dict

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

```

### multirun
```
$ python tests/test_apps/app_can_fail/my_app.py --multirun +divisor=1,0
[2021-03-23 15:25:08,182][HYDRA] Launching 2 jobs locally
[2021-03-23 15:25:08,182][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-23 15:25:08,291][HYDRA]        #1 : +divisor=0
Job failed.
Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

```

omry commented on hydra/core/utils.py:
```diff
@@ -166,14 +174,32 @@ def register(name: str, f: Any) -> None:
+        if self.job_result == JobResult.COMPLETED:
+            return self._return_value
+        else:
+            raise HydraJobException("Job failed.") from self._return_value
```
I feel "Job failed." is not super helpful.
How about:
```suggestion
            raise HydraJobException("Error executing job : {self.overrides}") from self._return_value
```

omry replied:
What does it look like if you use the regular HydraException instead of your special one?

omry commented on hydra/_internal/utils.py:
```diff
@@ -212,6 +213,9 @@ def run_and_report(func: Any) -> Any:
         if _is_env_set("HYDRA_FULL_ERROR") or is_under_debugger():
             raise ex
         else:
+            if isinstance(ex, HydraJobException):
+                sys.stderr.write(str(ex) + os.linesep)
+                ex = ex.__cause__  # type: ignore
```
I would prefer to keep this generic. Is there a way around changing this?

omry commented on tests/test_hydra.py:
```diff
@@ -1052,6 +1052,7 @@ def test_app_with_error_exception_sanitized(tmpdir: Any, monkeypatch: Any) -> No
     ]
     expected = dedent(
         """\
+        Job failed.
```
Was better without "Job failed.".

jieru-hu commented:
Sorry @omry  had to rebase master again due to test failures, all the new changes starts from f6b84d7

> I feel "Job failed." is not super helpful.
How about:

True, added more info to the exception message.

>What does it look like if you use the regular HydraException instead of your special one?

It does not look good.
#### multirun 
```commandline
$ python tests/test_apps/app_can_fail/my_app.py -m +divisor=1,0
[2021-03-23 13:52:26,537][HYDRA] Launching 2 jobs locally
[2021-03-23 13:52:26,538][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-23 13:52:26,663][HYDRA]        #1 : +divisor=0
Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 132, in run_job
    ret.return_value = task_function(task_cfg)
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 373, in <lambda>
    lambda: hydra.multirun(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/hydra.py", line 125, in multirun
    ret = sweeper.sweep(arguments=task_overrides)
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/core_plugins/basic_sweeper.py", line 158, in sweep
    _ = r.return_value
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 198, in return_value
    raise HydraJobException("Job failed.") from self._return_value
hydra.errors.HydraJobException: Job failed.
```
#### run
```commandline
$ python tests/test_apps/app_with_runtime_config_error/my_app.py 
Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 132, in run_job
    ret.return_value = task_function(task_cfg)
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_with_runtime_config_error/my_app.py", line 13, in my_app
    foo(cfg)
  File "/Users/jieru/workspace/hydra-fork/hydra/tests/test_apps/app_with_runtime_config_error/my_app.py", line 8, in foo
    cfg.foo = "bar"  # does not exist in the config
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/dictconfig.py", line 334, in __setattr__
    raise e
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/dictconfig.py", line 331, in __setattr__
    self.__set_impl(key, value)
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/dictconfig.py", line 315, in __set_impl
    self._set_item_impl(key, value)
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/basecontainer.py", line 518, in _set_item_impl
    self._validate_set(key, value)
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/dictconfig.py", line 177, in _validate_set
    target = self._get_node(key) if key is not None else self
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/dictconfig.py", line 442, in _get_node
    self._validate_get(key)
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/dictconfig.py", line 163, in _validate_get
    self._format_and_raise(
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/base.py", line 190, in _format_and_raise
    format_and_raise(
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/_utils.py", line 676, in format_and_raise
    _raise(ex, cause)
  File "/Users/jieru/opt/anaconda3/envs/pytest39/lib/python3.9/site-packages/omegaconf/_utils.py", line 574, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
omegaconf.errors.ConfigAttributeError: Key 'foo' is not in struct
    full_key: foo
    object_type=dict

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 365, in <lambda>
    lambda: hydra.run(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/hydra.py", line 102, in run
    _ = ret.return_value
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 198, in return_value
    raise HydraJobException("Job failed.") from self._return_value
hydra.errors.HydraJobException: Job failed.
```

> I would prefer to keep this generic. Is there a way around changing this?

I haven't figured out a better way. 
I feel it can be considered generic - everything thrown from 
hydra.run or hydra.multirun is wrapped in this exception.


> Was better without "Job failed.".

Agreed. Fixed the messages.

omry commented:
>> I would prefer to keep this generic. Is there a way around changing this?

> I haven't figured out a better way.
I feel it can be considered generic - everything thrown from
hydra.run or hydra.multirun is wrapped in this exception.

Can you make it work without changes to run_and_report? (That's the part about it not being generic).

jieru-hu commented:
> Can you make it work without changes to run_and_report? (That's the part about it not being generic).

How about when we run `hydra.run`, we catch `HydraJobException` and throw the cause?

### changes in code
#### in hydra.run
```
        # access the result to trigger an exception in case the job failed.
        try:
            _ = ret.return_value
        except HydraJobException as e:
            raise e.__cause__
```

### resulting exception
#### RUN
```
$ python tests/test_apps/app_with_runtime_config_error/my_app.py 
Traceback (most recent call last):
  File "tests/test_apps/app_with_runtime_config_error/my_app.py", line 13, in my_app
    foo(cfg)
  File "tests/test_apps/app_with_runtime_config_error/my_app.py", line 8, in foo
    cfg.foo = "bar"  # does not exist in the config
omegaconf.errors.ConfigAttributeError: Key 'foo' is not in struct
    full_key: foo
    object_type=dict

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

#### MULTIRUN
```
$ python tests/test_apps/app_can_fail/my_app.py -m +divisor=1,0
[2021-03-24 16:04:14,649][HYDRA] Launching 2 jobs locally
[2021-03-24 16:04:14,649][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-24 16:04:14,726][HYDRA]        #1 : +divisor=0
Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 132, in run_job
    ret.return_value = task_function(task_cfg)
  File "tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 374, in <lambda>
    lambda: hydra.multirun(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/hydra.py", line 129, in multirun
    return sweeper.sweep(arguments=task_overrides)
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/core_plugins/basic_sweeper.py", line 158, in sweep
    _ = r.return_value
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 199, in return_value
    raise HydraJobException(
hydra.errors.HydraJobException: Error executing job with overrides: ['+divisor=0']
```

omry commented:
Can you try to incorporate the changes from #1492 and see if it works for you (and does not seem to break anything else?)

jieru-hu commented:
> Can you try to incorporate the changes from #1492 and see if it works for you (and does not seem to break anything else?)

works!!
```
$ python tests/test_apps/app_with_runtime_config_error/my_app.py
Error executing job with overrides: []
Traceback (most recent call last):
  File "tests/test_apps/app_with_runtime_config_error/my_app.py", line 13, in my_app
    foo(cfg)
  File "tests/test_apps/app_with_runtime_config_error/my_app.py", line 8, in foo
    cfg.foo = "bar"  # does not exist in the config
omegaconf.errors.ConfigAttributeError: Key 'foo' is not in struct
    full_key: foo
    object_type=dict

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.


$ python tests/test_apps/app_can_fail/my_app.py -m +divisor=1,0
[2021-03-25 10:28:01,463][HYDRA] Launching 2 jobs locally
[2021-03-25 10:28:01,463][HYDRA]        #0 : +divisor=1
val=1.0
[2021-03-25 10:28:01,535][HYDRA]        #1 : +divisor=0
Error executing job with overrides: ['+divisor=0']
Traceback (most recent call last):
  File "tests/test_apps/app_can_fail/my_app.py", line 9, in my_app
    val = 1 / cfg.divisor
ZeroDivisionError: division by zero

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

```

jieru-hu commented:
force push due to rebase from master to pick up omry's change, changes start from 50b914f86986cc3398586f08e712f5f3711a36bf

omry commented on hydra/core/utils.py:
```diff
@@ -167,14 +175,34 @@ def setup_globals() -> None:
     hydra_cfg: Optional[DictConfig] = None
     working_dir: Optional[str] = None
     task_name: Optional[str] = None
+    job_result: Optional[JobResult] = None
```
```suggestion
    status: JobStatus = JobStatus.UNKNOWN
```

omry commented on hydra/core/utils.py:
```diff
@@ -167,14 +175,34 @@ def setup_globals() -> None:
     )
 
 
+class JobResult(Enum):
+    COMPLETED = 1
+    FAILED = 2
```
```suggestion
class JobStatus(Enum):
    UNKNOWN = 0
    COMPLETED = 1
    FAILED = 2
```

omry reviewed (commented):
Getting close.

omry reviewed (approved):
LGTM, thanks.

@jrapin, we are going to merge it shortly, we would still appreciate any feedback you have on this.

jrapin commented:
> LGTM, thanks.
> 
> @jrapin, we are going to merge it shortly, we would still appreciate any feedback you have on this.

LGTM as well, I'll give nevergrad sweeper a go with this new feature (or maybe I should wait for it to be integrated in a release?)

omry commented:
I recently released 1.1.0dev5, I am not sure if it's there or not. 
In any case, go ahead and do it against master, we can chat about how to publish it when it's ready.

omry commented:
Side note:
Take a look at #1503, It adds the ability to configure Samplers to Optuna Sweeper, leveraging recursive default lists.
Might want to take advantage of this to configure optimizers and other things.

jieru-hu commented:
Thanks @jrapin. I've created https://github.com/facebookresearch/hydra/issues/1511 and assigned it to you :)
